src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -141,19 +146,22 @@ private void ClearAllDiagnostics(ImmutableArray<StateSet> stateSets, ProjectId p
                     // PERF: don't fire events for ones that we dont have any diagnostics on
                     if (documentSet.Count > 0)
                     {
-                        RaiseProjectDiagnosticsRemoved(stateSet, projectId, documentSet, handleActiveFile: true, raiseEvents);
+                        AddProjectDiagnosticsRemovedArgs(ref argsBuilder.AsRef(), stateSet, projectId, documentSet, handleActiveFile: true);
                         documentSet.Clear();
                     }
                 }
+
+                raiseEvents(argsBuilder.ToImmutableAndClear());
```

Comment:
same idea bout checking count..

Reply:
Yes, this needs to be consistently checked at all call sites that raise events and the count of the array needs to be asserted or handled in all the event handlers

Reply:
➡️ I moved the check into `raiseEvents` to avoid the need to check it at call sites.

Reply:
➡️ Verified that `raiseEvents` checks the count